perf: refactor ServiceFactory to use constant service names

jahtalab opened the pull request:
Fixes #238

The gzipped bundle size is reduced from 16.78 KB on master to 16.65 KB.

vigneshshanmugam commented on packages/rum-core/src/common/service-factory.js:
```diff
@@ -73,27 +71,27 @@ class ServiceFactory {
+      if (typeof this._creators[name] === 'function') {
+        this._instances[name] = this._creators[name](this)
       } else {
         throw new Error('Can not get service, No creator for: ' + name)
```
It good to not throw here, instead log as error in dev mode.

jahtalab commented:
@vigneshshanmugam , 

Re No. 1, Separating the modules like ErrorLogging and PerformanceMonitoring helps with the ability to accommodate requests like https://github.com/elastic/apm-agent-rum-js/issues/755. Having said that, we can still reduce the overhead by using service factory objects directly or other similar ideas. let's discuss it further on the sync.

Re. 2, I think the trade-off is between executing JS during load time or during init time, the lazying init with it's current form optimises for running most/all of the work in the init time. But we discuss further different ways to resolve the trade-off.

jahtalab commented:
We had a discussion:
- Refactor `registerServiceCreator`
- eliminate apm-server init

vigneshshanmugam commented on packages/rum-core/src/common/service-factory.js:
```diff
@@ -32,22 +32,23 @@ import {
 
 class ServiceFactory {
   constructor() {
-    this._creators = {
```
Having the creators here is actually better and we are modifying the creators in another files seems like a sideeffect. Lets use the same patter and instead reduce the naming of instance methods to something like `addCreator` and `addInstance` instead of big `registerServiceInstance` etc.

jahtalab replied:
I compared the size reduction for having creators on the class, it nearly had 0.06 KB reduction in size, so I kept it this way. 

You're right about the side-effect but I think we probably should accept some side-effect in favour of the bundle size reduction.

vigneshshanmugam commented on packages/rum-core/src/error-logging/index.js:
```diff
@@ -25,14 +25,17 @@
+      CONFIG_SERVICE,
+      'TransactionService'
+    ])
+    return new ErrorLogging(...deps)
```
Just check how spread `...deps` is transpiled - might be a reason why you are seeing a increase in bundle size.

vigneshshanmugam commented on packages/rum-core/src/common/service-factory.js:
```diff
@@ -75,24 +76,25 @@ class ServiceFactory {
+        }
       }
+      return this._instances[name]
+    } else if (name.length) {
```
Lets use Array.isArray.

apmmachine commented:
## :broken_heart: Build Failed 

<!-- BUILD BADGES-->
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Fapm-agent-rum-mbp%2FPR-764/detail/PR-764/2//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Fapm-agent-rum-mbp%2FPR-764/detail/PR-764/2//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Fapm-agent-rum-mbp%2FPR-764/detail/PR-764/2//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-rum%2Fapm-agent-rum-mbp%2FPR-764/detail/PR-764/2//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://apm-agent-rum-js_764.docs-preview.app.elstc.co/diff) [![stats](https://img.shields.io/badge/stats-stats%20-blue)](https://ela.st/observabtl-ci-stats)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: [Pull request #764 updated]
* Start Time: 2020-04-28T13:40:24.907+0000

* Duration: 1390 min 56 sec (83396276)


* Commit: c9d94a445ddabea47dba6d57c9de216f449fedd7



#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 806  |
| Skipped      | 8 |
| Total        | 814   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY-->


### Steps errors

  <details><summary>Expand to view the steps failures</summary>
  <p>
  
  * **Name**: Bundlesize
    * Description: #!/bin/bash      set -o pipefail      npm run bundlesize|tee bundlesize.txt    
    * Result: FAILURE
    
    * Duration: 1 min 31 sec<
    
    * Start Time: 2020-04-28T13:44:56.176+0000
    
  
  * **Name**: Start Elastic Stack 8.0.0-SNAPSHOT - @elastic/apm-rum-core - none
    * Description: 
    * Result: FAILURE
    
    * Duration: 4 min 8 sec<
    
    * Start Time: 2020-04-28T13:53:50.902+0000
    
  
  * **Name**: Start Elastic Stack 8.0.0-SNAPSHOT - @elastic/apm-rum-core - none
    * Description: 
    * Result: FAILURE
    
    * Duration: 1 min 21 sec<
    
    * Start Time: 2020-04-28T13:58:10.910+0000
    
  
  * **Name**: Start Elastic Stack 7.0.0 - @elastic/apm-rum-react - none
    * Description: 
    * Result: FAILURE
    
    * Duration: 4 min 18 sec<
    
    * Start Time: 2020-04-28T13:53:56.148+0000
    
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2020-04-28T14:01:23.520Z] node-puppeteer         |     ? should sendPageLoadMetrics
[2020-04-28T14:01:23.520Z] node-puppeteer         |     ? should filter out empty transactions
[2020-04-28T14:01:23.520Z] node-puppeteer         |     ? should filter managed transactions with duration above threshold
[2020-04-28T14:01:23.520Z] node-puppeteer         |     ? should correctly use xhr patch
[2020-04-28T14:01:23.520Z] node-puppeteer         |     ? should consider fetchInProgress to avoid duplicate spans
[2020-04-28T14:01:23.520Z] node-puppeteer         |     ? should create fetch spans
[2020-04-28T14:01:23.520Z] node-puppeteer         |     ? should redact auth from xhr tasks
[2020-04-28T14:01:23.520Z] node-puppeteer         |     ? should not duplicate xhr spans if fetch is a polyfill
[2020-04-28T14:01:23.520Z] node-puppeteer         |     ? should add xhr tasks
[2020-04-28T14:01:23.520Z] node-puppeteer         |     ? should create Transactions on history.pushState
[2020-04-28T14:01:23.520Z] node-puppeteer         |     ? should subscribe to events based on instrumentation flags
[2020-04-28T14:01:23.521Z] node-puppeteer         |     ? should create http-request transaction if no current transaction exist
[2020-04-28T14:01:23.521Z] node-puppeteer         |     ? should include multiple XHRs in the same transaction
[2020-04-28T14:01:23.521Z] node-puppeteer         |     ? should send span context destination details to apm-server
[2020-04-28T14:01:23.521Z] node-puppeteer         |     ? should filter only sampled transaction without spans
[2020-04-28T14:01:23.521Z] node-puppeteer         |     ? should create click transactions
[2020-04-28T14:01:23.521Z] node-puppeteer         |     ? should respect the transaction type priority order
[2020-04-28T14:01:23.521Z] node-puppeteer         |     PerformanceMonitoring Utils
[2020-04-28T14:01:23.521Z] node-puppeteer         |       ? should group small continuously similar spans up until the last one
[2020-04-28T14:01:23.521Z] node-puppeteer         |       ? should group small continuously similar spans
[2020-04-28T14:01:23.521Z] node-puppeteer         |       ? should calculate browser responsiveness
[2020-04-28T14:01:23.521Z] node-puppeteer         |       ? should reset spans for unsampled transactions
[2020-04-28T14:01:23.786Z] node-puppeteer         |   SpanBase
[2020-04-28T14:01:23.786Z] node-puppeteer         |     ? should add default name and type for span
[2020-04-28T14:01:23.786Z] node-puppeteer         |     ? should addLabels
[2020-04-28T14:01:23.786Z] node-puppeteer         |     ? should addContext
[2020-04-28T14:01:23.786Z] node-puppeteer         |   Span
[2020-04-28T14:01:23.786Z] node-puppeteer         |     ? should return null for duration if not ended
[2020-04-28T14:01:23.786Z] node-puppeteer         |     ? should support dot delimiter in span types
[2020-04-28T14:01:23.786Z] node-puppeteer         |   TransactionService
[2020-04-28T14:01:23.786Z] node-puppeteer         |     ? should not start span when there is no current transaction
[2020-04-28T14:01:23.786Z] node-puppeteer         |     ? should call startSpan on current Transaction
[2020-04-28T14:01:23.786Z] node-puppeteer         |     ? should start transaction
[2020-04-28T14:01:23.786Z] node-puppeteer         |     ? should create a reusable transaction on the first span
[2020-04-28T14:01:23.786Z] node-puppeteer         |     ? should fire onstart hook only once for a transaction
[2020-04-28T14:01:23.787Z] node-puppeteer         |     ? should capture page load on first transaction
[2020-04-28T14:01:23.787Z] node-puppeteer         |     ? should not capture timings as spans for unsampled transactions
[2020-04-28T14:01:23.787Z] node-puppeteer         |     ? should reuse Transaction
[2020-04-28T14:01:23.787Z] node-puppeteer         |     ? should not capture resource/user spans or marks for custom transaction
[2020-04-28T14:01:23.787Z] node-puppeteer         |     ? should use initial page load name before ending the transaction
[2020-04-28T14:01:23.787Z] node-puppeteer         |     ? should capture resources from navigation timing
[2020-04-28T14:01:23.787Z] node-puppeteer         |     ? should ignore transactions that match the list
[2020-04-28T14:01:23.787Z] node-puppeteer         |     ? should apply sampling to transactions
[2020-04-28T14:01:23.787Z] node-puppeteer         |     ? should consider page load traceId and spanId
[2020-04-28T14:01:23.787Z] node-puppeteer         |     ? should ensureCurrentTransaction once per startTransaction
[2020-04-28T14:01:23.787Z] node-puppeteer         |     ? should include size & server timing in page load context
[2020-04-28T14:01:23.787Z] node-puppeteer         |     ? should not capture breakdown metrics by default
[2020-04-28T14:01:23.787Z] node-puppeteer         |     ? should create unmanaged transactions by default
[2020-04-28T14:01:23.788Z] node-puppeteer         |     ? should create managed transactions if the managed option is provided
[2020-04-28T14:01:23.788Z] node-puppeteer         |     ? should not produce negative durations while adjusting to the spans
[2020-04-28T14:01:23.788Z] node-puppeteer         |     ? should adjust transaction start based on earliest span start
[2020-04-28T14:01:23.788Z] node-puppeteer         |     ? should adjust transaction end based on latest span end
[2020-04-28T14:01:23.788Z] node-puppeteer         |     ? should truncate active spans after transaction ends
[2020-04-28T14:01:23.788Z] node-puppeteer         |     ? should account for spans start > transaction start during breakdown
[2020-04-28T14:01:23.788Z] node-puppeteer         |     ? should start temporary transaction on startSpan
[2020-04-28T14:01:23.788Z] node-puppeteer         |     ? should only call setInterval once for current transaction
[2020-04-28T14:01:23.788Z] node-puppeteer         |     ? should redefine type based on the defined order
[2020-04-28T14:01:23.788Z] node-puppeteer         |     performance entry recorder
[2020-04-28T14:01:23.788Z] node-puppeteer         |       ? should start/stop performance recorder for managed transaction
[2020-04-28T14:01:23.788Z] node-puppeteer         |       ? should start recorder only once during redefining
[2020-04-28T14:01:23.788Z] node-puppeteer         |       ? should stop recorder before starting a non-reusable transaction
[2020-04-28T14:01:23.789Z] node-puppeteer         |       ? should not record longtasks when monitorLongtasks is false
[2020-04-28T14:01:23.789Z] node-puppeteer         |   transaction.Transaction
[2020-04-28T14:01:23.789Z] node-puppeteer         |     ? should contain correct number of spans in the end
[2020-04-28T14:01:23.789Z] node-puppeteer         |     ? should not generate stacktrace if the option is not passed
[2020-04-28T14:01:23.789Z] node-puppeteer         |     ? should redefine transaction
[2020-04-28T14:01:23.789Z] node-puppeteer         |     ? should add and remove tasks
[2020-04-28T14:01:23.789Z] node-puppeteer         |     ? should mark events
[2020-04-28T14:01:23.789Z] node-puppeteer         |     ? should not start spans after end
[2020-04-28T14:01:23.789Z] node-puppeteer         |     ? should always set span parentId
[2020-04-28T14:01:23.789Z] node-puppeteer         |     ? should calculate reusability
[2020-04-28T14:01:23.789Z] node-puppeteer         | HeadlessChrome 84.0.4122 (Linux 0.0.0): Executed 233 of 235 (skipped 2) SUCCESS (2.769 secs / 2.557 secs)
[2020-04-28T14:01:23.789Z] node-puppeteer         | TOTAL: 233 SUCCESS
[2020-04-28T14:01:23.789Z] node-puppeteer         | Running cleanups: 1
[2020-04-28T14:01:23.790Z] node-puppeteer         | lerna success run Ran npm script 'test' in 1 package in 13.5s:
[2020-04-28T14:01:23.790Z] node-puppeteer         | lerna success - @elastic/apm-rum-core
[2020-04-28T14:01:24.057Z] node-puppeteer exited with code 0
[2020-04-28T14:01:24.320Z] Aborting on container exit...
[2020-04-28T14:01:24.320Z] + echo 'Exit code from docker-compose 0'
[2020-04-28T14:01:24.320Z] Exit code from docker-compose 0
[2020-04-28T14:01:28.558Z] Post stage
[2020-04-28T14:01:28.609Z] Recording test results
[2020-04-28T14:01:29.906Z] Archiving artifacts
[2020-04-29T12:50:16.012Z] Failed in branch Matrix - STACK_VERSION = '7.0.0', SCOPE = '@elastic/apm-rum'
[2020-04-29T12:50:16.200Z] Stage "Stack 8.0.0-SNAPSHOT SauceLabs" skipped due to earlier failure(s)
[2020-04-29T12:50:16.308Z] Stage "Stack 8.0.0-SNAPSHOT SauceLabs PR" skipped due to earlier failure(s)
[2020-04-29T12:50:16.425Z] Stage "Integration Tests" skipped due to earlier failure(s)
[2020-04-29T12:50:16.521Z] Stage "Benchmarks" skipped due to earlier failure(s)
[2020-04-29T12:50:16.602Z] Stage "Coverage" skipped due to earlier failure(s)
[2020-04-29T12:50:16.669Z] Stage "Release" skipped due to earlier failure(s)
[2020-04-29T12:50:16.697Z] Stage "Release" skipped due to earlier failure(s)
[2020-04-29T12:50:16.760Z] Stage "Release" skipped due to earlier failure(s)
[2020-04-29T12:50:16.864Z] Stage "Opbeans" skipped due to earlier failure(s)
[2020-04-29T12:50:18.076Z] Running on Jenkins in /var/lib/jenkins/workspace/ent-rum_apm-agent-rum-mbp_PR-764
[2020-04-29T12:50:18.430Z] [INFO] getVaultSecret: Getting secrets
[2020-04-29T12:50:18.540Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2020-04-29T12:50:21.457Z] + chmod 755 generate-build-data.sh
[2020-04-29T12:50:21.457Z] + ./generate-build-data.sh https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-rum/apm-agent-rum-mbp/PR-764/ https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-rum/apm-agent-rum-mbp/PR-764/runs/2 ABORTED 83396276
[2020-04-29T12:50:22.195Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-rum/apm-agent-rum-mbp/PR-764/runs/2/steps/?limit=10000 -o steps-info.json
[2020-04-29T12:50:22.195Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-rum/apm-agent-rum-mbp/PR-764/runs/2/tests/?status=FAILED -o tests-errors.json

```

</p>
</details>



<!--PIPELINE
{"commit":"c9d94a445ddabea47dba6d57c9de216f449fedd7","number":"2","status":"ABORTED","url":"https://apm-ci.elastic.co/job/apm-agent-rum/job/apm-agent-rum-mbp/job/PR-764/2/"}
PIPELINE-->

vigneshshanmugam commented on packages/rum-angular/test/index.ts:
```diff
@@ -36,7 +36,7 @@ export function initializeApmService(service, config) {
   config.serverUrl = serverUrl
   const serverMock = new ApmServerMock(apmServer, globalConfig.useMocks)
-  apmBase.serviceFactory.registerServiceInstance('ApmServer', serverMock)
+  apmBase.serviceFactory.instances['ApmServer'] = serverMock
```
can you import and use service constants?

vigneshshanmugam replied:
can you import and use service constants?

vigneshshanmugam commented on packages/rum-core/src/common/service-factory.js:
```diff
@@ -73,28 +74,24 @@ class ServiceFactory {
+        if (typeof serviceCreators[name] === 'function') {
+          this.instances[name] = serviceCreators[name](this)
+        } else if (__DEV__) {
+          console.log('Can not get service, No creator for: ' + name)
```
nit: 'Cannot'

vigneshshanmugam commented on packages/rum-core/src/error-logging/index.js:
```diff
@@ -24,15 +24,22 @@
   }
 }
+
+export { ErrorLogging, registerServices }
```
Can we remove the `ErrorLogging` export as it's not used?

vigneshshanmugam commented on packages/rum-core/src/index.js:
```diff
@@ -25,25 +25,30 @@
 import afterFrame from './common/after-frame'
 import { createTracer } from './opentracing'
 
+registerPerfServices()
```
this feels like Side Effect. can we move this inside createServiceFactory itself? Atleast we know we register only when someone creates the factory.

vigneshshanmugam commented on packages/rum-core/src/performance-monitoring/index.js:
```diff
@@ -25,27 +25,42 @@
-      )
-    })
+function registerServices() {
+  serviceCreators['TransactionService'] = serviceFactory => {
```
Any reason why we didn't have TransactionService in constant?

vigneshshanmugam commented on packages/rum-core/src/performance-monitoring/index.js:
```diff
@@ -25,27 +25,42 @@
   }
 }
+
+export { PerformanceMonitoring, registerServices }
```
remove PerformanceMonitoring export as well?

vigneshshanmugam commented on packages/rum-core/src/common/service-factory.js:
```diff
@@ -26,42 +26,43 @@
+} from './constants'
+import { __DEV__ } from '../env'
+
+const serviceCreators = {
```
This is exactly similar to file level instance. Remember you were suggesting aganist this idea? This would mean we have the same factory instance for all the tests? Previously `createServiceFactory` would give us a fresh instance every time.

jahtalab replied:
Although, this is a file level object, it's not the same as I was suggesting.

The file level instance that I don't like is the one that contains our service instances, imagine if `ServiceFactory.instances` was a file level object, this is not a good pattern.

But serviceCreators are only functions that create the services, the only way this file level object can make problems is as if we start changing it during our tests.

jahtalab commented:
@vigneshshanmugam , I've rebased and addressed some of the comments. 

Re. using constant service names, actually don't have much effect on the gzipped files, so I don't think it makes sense to use them everywhere, but if there're multiple of the same service in a file that could be a good place to use constants.
